vumi/transports/turn/turn.py

```diff
@@ -0,0 +1,205 @@
+
+from vumi.utils import http_request_full, normalize_msisdn, LogFilterSite
+from vumi.transports.base import Transport
+from vumi.transports.failures import TemporaryFailure, PermanentFailure
+from vumi.errors import VumiError
+
+
+def iso8601(turn_timestamp):
+    # TODO: this will be a unix timestamp
+    if turn_timestamp:
+        ts = datetime.strptime(turn_timestamp, '%Y.%m.%d %H:%M:%S')
+        return ts.isoformat()
+    else:
+        return ''
+
+
+def normalize_outbound_msisdn(msisdn):
+    if msisdn.startswith('+'):
+        return msisdn.replace('+', '')
+    else:
+        return msisdn
+
+class ReceiveEventResource(Resource):
+    isLeaf = True
+
+    def __init__(self, config, publish_event_func, publish_message):
+        self.config = config
+        self.publish_event_func = publish_event_func
+        self.publish_inbound_func = publish_message
+        self.transport_name = 'turn'
+
+    @inlineCallbacks
+    def do_render(self, request):
+        request.setResponseCode(http.CREATED)
+        # request.setHeader('Content-Type', 'text/plain')
+        try:
+            # TODO: validate HMAC secret
+
+            for message in request.args.get("messages", []):
+                message_id = '%s.%s' % (self.transport_name, message["id"])
+                content = ''
+                if message['type'] == 'text':
+                    content = message["text"]["body"]
+                elif message['type'] == 'location':
+                    loc = message["location"]
+                    content = 'geo:{},{}'.format(loc['latitude'], loc['longitude'])
+
+                yield self.publish_inbound_func(
+                    transport_name=self.transport_name,
+                    transport_type='sms',
+                    message_id=message_id,
+                    transport_metadata={
+                        'original_message_id': message_id,
+                        'timestamp': iso8601(message["timestamp"]),
+                        'network_id': request.args['provider'][0],
+                        'keyword': request.args['keyword'][0],
+                        },
+                    to_addr=normalize_msisdn(self.config['to_addr']),
+                    from_addr=normalize_msisdn(message['from']),
+                    content=content,
+                    )
+                log.msg("Inbound Enqueued.")
+
+            for event in request.args.get("statuses", []):
+                message_id = '%s.%s' % (self.transport_name, event['id'])
+
+                if event["status"] == 'sent':
+                    delivery_status = 'pending'
+                elif event["status"] == 'failed':
+                    delivery_status = 'failed'
+                elif status == 'delivered':
+                    delivery_status = 'delivered'
+                else:
+                    continue
+
+                yield self.publish_event_func(
+                    user_message_id=message_id,
```

Comment:
This should be the ID of the message that this event is for, not the ID of the event.